### PR TITLE
[HOPSWORKS-2890] Add GCS connector

### PR DESCRIPTION
**Pull request by DhananjayMukhedkar**
Changes to add GCS storage connector

### CODE REVIEW

File: java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java [hunk 1]
```diff
@@ -608,4 +611,58 @@ public String addFile(String filePath) {
...
+    }
+
+    if (!Strings.isNullOrEmpty(storageConnector.getAlgorithm())) {
+      sparkSession.sparkContext().hadoopConfiguration().set(
+          property_algorithm, storageConnector.getAlgorithm());
+    } else {
+      sparkSession.sparkContext().hadoopConfiguration().unset(property_algorithm);
+    }
+
+
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java:666]: unnessecary whitespace
```suggestion
```

File: java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java [hunk 2]
```diff
@@ -608,4 +611,58 @@ public String addFile(String filePath) {
...
+    );
+    // Whether to use a service account for GCS authorization. Setting this
+    // property to `false` will disable use of service accounts for authentication.
+    sparkSession.sparkContext().hadoopConfiguration().set(
+        property_gcs_account_enable, "true"
+    );
+    // The JSON key file of the service account used for GCS
+    // access when google.cloud.auth.service.account.enable is true.
+    if (!Strings.isNullOrEmpty(storageConnector.getKeyPath())) {
+      String localPath = addFile(storageConnector.getKeyPath().replace("hdfs://", ""));
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java:637]: With my changes in the backend, this shouldn't be necessary

File: java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java [hunk 3]
```diff
@@ -608,4 +611,58 @@ public String addFile(String filePath) {
...
+      sparkSession.sparkContext().hadoopConfiguration().set(
+          property_key_file, localPath
+      );
+    } else {
+      sparkSession.sparkContext().hadoopConfiguration().unset(property_key_file); // to clear previous
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java:642]:
```suggestion
      // to clear previous
      sparkSession.sparkContext().hadoopConfiguration().unset(property_key_file);
```

File: python/hsfs/storage_connector.py [hunk 4]
```diff
@@ -927,3 +928,97 @@ def read_stream(
...
+        return self._encryption_key_hash
+
+    def spark_options(self):
+        """Return prepared options to be passed to Spark, based on the additional
+        arguments.
+        """
+        return super().spark_options
+
+    def _get_path(self, sub_path: str):
+        super()._get_path(sub_path)
```
Comment (moritzmeister) [inline python/hsfs/storage_connector.py:982]: not sure why you need this, it should inherit the method from the base class already.

File: python/hsfs/storage_connector.py [hunk 5]
```diff
@@ -927,3 +928,97 @@ def read_stream(
...
+    @property
+    def encryption_key_hash(self):
+        """Encryption Key Hash"""
+        return self._encryption_key_hash
+
+    def spark_options(self):
+        """Return prepared options to be passed to Spark, based on the additional
+        arguments.
+        """
+        return super().spark_options
```
Comment (moritzmeister) [inline python/hsfs/storage_connector.py:979]: not sure why you need this, it should inherit the method from the base class already.

**Review by moritzmeister — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/logicalclocks/hsfs/StorageConnector.java [hunk 6]
```diff
@@ -499,4 +500,53 @@ public String getPath(String subPath) {
...
+    public Map<String, String> sparkOptions() {
+      return new HashMap();
+    }
+
+    public Dataset<Row> read(String dataFormat, Map<String, String> options, String path)
+        throws FeatureStoreException, IOException {
+
+      this.refetch();
+      return SparkEngine.getInstance().read(this, dataFormat, options, path);
+    }
```
[the remark this answers is not in the document]
Comment (DhananjayMukhedkar) [reply java/src/main/java/com/logicalclocks/hsfs/StorageConnector.java]: using this method now to call get_path to format the final path.

File: python/hsfs/storage_connector.py (hunk 4 above)
Comment (DhananjayMukhedkar) [reply python/hsfs/storage_connector.py]: added now to format the final path argument. This method was not being invoked so called in explicitly within read method.

File: python/hsfs/storage_connector.py (hunk 5 above)
Comment (DhananjayMukhedkar) [reply python/hsfs/storage_connector.py]: I was getting exceptions after removing this method so added it to return a blank dictionary

**Review by moritzmeister — APPROVED** [review]

**Pull request merged**